Analysis/EventFiltering/centralEventFilterProcessor.cxx

```diff
@@ -0,0 +1,185 @@
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   centralEventFilterProcessor.cxx
+
+#include "centralEventFilterProcessor.h"
+
+#include "Framework/ControlService.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/TableConsumer.h"
+
+#include <TFile.h>
+#include <TH1D.h>
+
+#include <iostream>
+#include <cstdio>
+#include <fmt/format.h>
+#include <rapidjson/document.h>
+#include <rapidjson/filereadstream.h>
+
+using namespace o2::framework;
+using namespace rapidjson;
+
+namespace
+{
+Document readJsonFile(std::string& config)
+{
+  FILE* fp = fopen(config.data(), "rb");
+
+  char readBuffer[65536];
+  FileReadStream is(fp, readBuffer, sizeof(readBuffer));
+
+  Document d;
+  d.ParseStream(is);
+  fclose(fp);
+  return d;
+}
+} // namespace
+
+namespace o2::aod::filtering
+{
+
+void CentralEventFilterProcessor::init(framework::InitContext& ic)
+{
+  // JSON example
+  // {
+  //   "subwagon_name" : "CentralEventFilterProcessor",
+  //   "configuration" : {
+  //     "NucleiFilters" : {
+  //       "H2" : 0.1,
+  //       "H3" : 0.3,
+  //       "HE3" : 1.,
+  //       "HE4" : 1.
+  //     }
+  //   }
+  // }
+  std::cout << "Start init" << std::endl;
+  Document d = readJsonFile(mConfigFile);
+  int nCols{0};
+  for (auto& workflow : d["workflows"].GetArray()) {
+    if (std::string_view(workflow["subwagon_name"].GetString()) == "CentralEventFilterProcessor") {
+      auto& config = workflow["configuration"];
+      for (auto& filter : AvailableFilters) {
+        auto& filterConfig = config[filter];
+        if (config.IsObject()) {
+          std::unordered_map<std::string, float> tableMap;
+          for (auto& node : filterConfig.GetObject()) {
+            tableMap[node.name.GetString()] = node.value.GetDouble();
+            nCols++;
+          }
+          mDownscaling[filter] = tableMap;
+        }
+      }
+      break;
+    }
+  }
+  std::cout << "Middle init" << std::endl;
+  mScalers = new TH1D("mScalers", ";;Number of events", nCols + 1, -0.5, 0.5 + nCols);
+  mScalers->GetXaxis()->SetBinLabel(1, "Total number of events");
+
+  mFiltered = new TH1D("mFiltered", ";;Number of filtered events", nCols + 1, -0.5, 0.5 + nCols);
+  mFiltered->GetXaxis()->SetBinLabel(1, "Total number of events");
+
+  int bin{2};
+  for (auto& table : mDownscaling) {
+    for (auto& column : table.second) {
+      mScalers->GetXaxis()->SetBinLabel(bin, column.first.data());
+      mFiltered->GetXaxis()->SetBinLabel(bin++, column.first.data());
+    }
+  }
+
+  TFile test("test.root", "recreate");
+  mScalers->Clone()->Write();
```

Comment:
This doesn't leak?

Reply:
It should not, as the clone is created in the directory of the file and when the file closes it should be deleted

Reply:
cling test:
```
root [0] TH1D h("h","h",1000,-10,10);
root [1] TFile test("test.root","recreate");
root [2] auto clone = h.Clone("hh");
root [3] clone->Write();
root [4] test.Close()
root [5] clone
(TObject *) @0x7ffe703c2560
root [6] clone->Dump()

 *** Break *** segmentation violation
```